api/traces/v1/http.go

```diff
@@ -163,6 +165,46 @@ func NewV2Handler(read *url.URL, opts ...HandlerOption) http.Handler {
 	return r
 }
 
+// Parse a URL; if the URL includes `{tenant}` that portion will be replaced by the tenant
+func ExpandTemplatedUpstream(templateUpstream, tenant string) (*url.URL, error) {
+	rawTracesReadEndpoint := strings.Replace(templateUpstream, "{tenant}", tenant, 1)
+	return url.ParseRequestURI(rawTracesReadEndpoint)
+}
+
+// middlewareSetTemplatedUpstream is a variation of proxy.MiddlewareSetUpstream()
+// with additional processing if the upstream includes "{tenant}"
+func middlewareSetUpstream(logger log.Logger, read *url.URL, readTemplate string) proxy.Middleware {
+	if read != nil {
+		return proxy.MiddlewareSetUpstream(read)
+	}
+
+	// Cache upstream URLs to avoid re-parse on every read
+	templateToURL := map[string]*url.URL{}
+
+	return func(r *http.Request) {
+		tenant, ok := authentication.GetTenant(r.Context())
+		if !ok {
+			// At this point another middleware must have put the tenant into the context
+			level.Debug(logger).Log("msg", "Internal error; expected tenant in request context")
+		}
+
+		upstream, ok := templateToURL[tenant]
+		if !ok {
+			upstream, err := ExpandTemplatedUpstream(readTemplate, tenant)
+			if err != nil {
+				// Log if the tenant label includes characters that can't appear in a hostname (such as punctuation)
+				level.Debug(logger).Log("msg", "Internal error; tenant contains characters that cannot appear in hostname")
+			}
+
+			templateToURL[tenant] = upstream
+		}
+
+		r.URL.Scheme = upstream.Scheme
+		r.URL.Host = upstream.Host
```

Comment:
We continue to proceed from the earlier error check so won't this panic if the URL is nil?

Reply:
I missed a `:=` instead of `=` when refactoring.  Fixed; I have added an e2e test for the template parameter.

Reply:
Nice